[quant][graphmode] Fold quantize op into module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25974 [quant][graphmode] use whitelist for selecting observed values
* **#25625 [quant][graphmode] Fold quantize op into module**
* #25624 [quant][graphmode] Add fusion for quantized linear
* #25623 Port fuse_linear from pytorch/tvm

Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
[quant][graphmode] Fold quantize op into module

Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

ZolotukhinM commented on torch/csrc/jit/passes/quantization.h:
```diff
@@ -96,5 +96,10 @@ TORCH_API void QuantFusion(std::shared_ptr<Graph>& graph);
  */
 TORCH_API void FoldConvBatchNorm2d(const script::Module& module);
 
+/** \brief Fold quantize function call into module
```
This comment doesn't really help: please explain what exactly we're folding, for what modules and in what calls. The name also might be more specific I think.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -828,5 +828,44 @@ graph(%self, %x):
+  for (const auto& match: matches) {
+    auto match_vmap = match.values_map;
+    auto* weight = match_vmap.at(vmap.at("weight"));
+    TORCH_INTERNAL_ASSERT(
```
You don't need to assert this - this is guaranteed by the API of subgraph matcher.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -828,5 +828,44 @@ graph(%self, %x):
+        weight->node()->s(attr::name) == "weight",
+        "Expected attribute name to be weight but got: ", weight->node()->s(attr::name));
+    auto float_weight = module.get_parameter("weight").variable_data();
+    auto scale = toIValue(match_vmap.at(vmap.at("scale"))).value().toDouble();
```
If some module happens to have "weight" but doesn't have "scale" this function will fail (imagine e.g. there is a conv that we're not quantizing). We should bail out gracefully in such cases.

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

jerryzh168 committed:
Update on "[quant][graphmode] Fold quantize op into module"


Summary:
We want to fold the quantize op for weights/bias into module to avoid quantizing weights on the fly.

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208889](https://our.internmc.facebook.com/intern/diff/D17208889)

ZolotukhinM reviewed (approved):
Looks good!
